### PR TITLE
Close span when Exception is ClientAbortException

### PR DESCRIPTION
**Pull request by mm23504570**
Fixes gh-1038

### CODE REVIEW

**Comment by pivotal-issuemaster** [comment]
@mm23504570 Please sign the [Contributor License Agreement](https://cla.pivotal.io/sign/spring?repositoryId=spring-cloud/spring-cloud-sleuth&pullRequestId=1040)!

[Click here](https://cla.pivotal.io/sync/spring?repositoryId=spring-cloud/spring-cloud-sleuth&pullRequestId=1040) to manually synchronize the status of this Pull Request.

See the [FAQ](https://cla.pivotal.io/about) for frequently asked questions.

**Comment by pivotal-issuemaster** [comment]
@mm23504570 Thank you for signing the [Contributor License Agreement](https://cla.pivotal.io/sign/spring?repositoryId=spring-cloud/spring-cloud-sleuth&pullRequestId=1040)!

File: spring-cloud-sleuth-core/src/main/java/org/springframework/cloud/sleuth/instrument/web/TraceFilter.java [hunk 1]
```diff
@@ -21,10 +21,7 @@
 import java.util.Collections;
 import java.util.Enumeration;
 import java.util.regex.Pattern;
-import javax.servlet.FilterChain;
-import javax.servlet.ServletException;
-import javax.servlet.ServletRequest;
-import javax.servlet.ServletResponse;
+import javax.servlet.*;
```
Comment (marcingrzejszczak) [inline spring-cloud-sleuth-core/src/main/java/org/springframework/cloud/sleuth/instrument/web/TraceFilter.java]: We prefer explicit import statements

File: spring-cloud-sleuth-core/src/main/java/org/springframework/cloud/sleuth/instrument/web/TraceFilter.java [hunk 2]
```diff
@@ -271,6 +273,13 @@ private void detachOrCloseSpans(HttpServletRequest request,
 		}
 	}
 
+	/**
+	 * check exception is org.apache.catalina.connector.ClientAbortException.
+	 */
+	private boolean isClientAbortExpcetion(Throwable exception) {
```
Comment (marcingrzejszczak) [inline spring-cloud-sleuth-core/src/main/java/org/springframework/cloud/sleuth/instrument/web/TraceFilter.java]: I don't like this. What if some other exceptions will be thrown? Will we have a list of exceptions for which we close the spans? I think we have to think of a different way of solving this problem.

Comment (adriancole) [reply spring-cloud-sleuth-core/src/main/java/org/springframework/cloud/sleuth/instrument/web/TraceFilter.java]: also typo. The javadoc should say why we are special-casing.. or a comment where we are special casing

Comment (mm23504570) [reply spring-cloud-sleuth-core/src/main/java/org/springframework/cloud/sleuth/instrument/web/TraceFilter.java]: @marcingrzejszczak I will create a class that holds a list of exceptions that let the user configure which exceptions should be closed in this case. 
What do you think?

**Comment by adriancole** [comment]
seems if something like this were the case, it could be auto-configured on
presence of the tomcat class in question? (ex undertow should not need to
be bothered...) This is ideals, not saying the solution needs to be ideal

**Comment by mm23504570** [comment]
Good. I like your solution, it is indeed more generic and graceful. I will try to implement it. Thanks.

File: spring-cloud-sleuth-core/src/main/java/org/springframework/cloud/sleuth/instrument/web/TraceFilter.java [hunk 3]
```diff
@@ -273,11 +272,14 @@ private void detachOrCloseSpans(HttpServletRequest request,
 		}
 	}
 
-	/**
-	 * check exception is org.apache.catalina.connector.ClientAbortException.
-	 */
-	private boolean isClientAbortExpcetion(Throwable exception) {
-		return exception.getClass().getName().equals("org.apache.catalina.connector.ClientAbortException");
+
+	private boolean needIgnoreExpcetion(Throwable exception) {
```
Comment (marcingrzejszczak) [inline spring-cloud-sleuth-core/src/main/java/org/springframework/cloud/sleuth/instrument/web/TraceFilter.java]: typo :(

File: spring-cloud-sleuth-core/src/test/java/org/springframework/cloud/sleuth/instrument/web/TraceFilterTests.java [hunk 4]
```diff
@@ -386,6 +392,9 @@ public void closesSpanWhenResponseStatusIs2xxAndExceptionIsClientAbortException(
 				.header(Span.TRACE_ID_NAME, 20L).buildRequest(new MockServletContext());
 		TraceFilter filter = new TraceFilter(beanFactory());
 		BDDMockito.given(beanFactory.getBean(ErrorController.class)).willReturn(() -> "/error");
+		List<ExceptionToIgnoreInTraceFilter> filters = Lists.newArrayList(getClientAbortExpcetionToIgnoreInTraceFilter());
```
Comment (marcingrzejszczak) [inline spring-cloud-sleuth-core/src/test/java/org/springframework/cloud/sleuth/instrument/web/TraceFilterTests.java]: I think we should add one test with a filter that throws the `ClientAbortException` and we would like to ensure that the span got closed and reported. WDYT?

Comment (mm23504570) [reply spring-cloud-sleuth-core/src/test/java/org/springframework/cloud/sleuth/instrument/web/TraceFilterTests.java]: like this?
``` java
@Test
public void closesSpanWhenResponseStatusIs2xxAndClientAbortExceptionThrowAfterTraceFilter() throws Exception {
	this.request = builder().header(Span.SPAN_ID_NAME, PARENT_ID)
			.header(Span.TRACE_ID_NAME, 20L).buildRequest(new MockServletContext());
	TraceFilter filter = new TraceFilter(beanFactory());
	BDDMockito.given(beanFactory.getBean(ErrorController.class)).willReturn(() -> "/error");
	List<ExceptionToIgnoreInTraceFilter> filters = Lists.newArrayList(getClientAbortExpcetionToIgnoreInTraceFilter());
	BDDMockito.given(beanFactory.getBean(ExceptionToIgnoreInTraceFilterProvider.class))
			.willReturn(getExceptionToIgnoreInTraceFilterProvider(filters));
	this.response = new MockHttpServletResponse();
	response.setStatus(200);
	this.filterChain = new MockFilterChain(){
		@Override
		public void doFilter(ServletRequest request, ServletResponse response) throws IOException, ServletException {
			throw new ClientAbortException();
		}
	};
	try {
		filter.doFilter(this.request, this.response, this.filterChain);
	}catch (ClientAbortException e){
		// ig
	}
	then(TestSpanContextHolder.getCurrentSpan()).isNull();
}
```

Comment (marcingrzejszczak) [reply spring-cloud-sleuth-core/src/test/java/org/springframework/cloud/sleuth/instrument/web/TraceFilterTests.java]: yeah and also

```java
	then(TestSpanContextHolder.getCurrentSpan()).isNull();
         // verify that the span got reported
         then(spanReporter.getSpans()).hasSize(1);
```

**Comment by marcingrzejszczak** [comment]
Congrats @mm23504570 on your contribution! Awesome work.

**Pull request merged**